joblib/test/test_memory.py

```diff
@@ -722,3 +724,51 @@ def test_memory_clear(tmpdir):
     memory.clear()
 
     assert os.listdir(memory.cachedir) == []
+
+
+def fast_func_with_complex_output():
+    complex_obj = ['a' * 1000] * 1000
+    return complex_obj
+
+
+def fast_func_with_conditional_complex_output(complex_output=True):
+    complex_obj = {str(i): i for i in range(int(1e5))}
+    return complex_obj if complex_output else 'simple output'
+
+
+@with_multiprocessing
+def test_cached_function_race_condition_when_persisting_output(tmpdir, capfd):
+    # Test race condition where multiple processes are writing into
+    # the same output.pkl. See
+    # https://github.com/joblib/joblib/issues/490 for more details.
+    memory = Memory(cachedir=tmpdir.strpath)
+    func_cached = memory.cache(fast_func_with_complex_output)
+
+    Parallel(n_jobs=2)(delayed(func_cached)() for i in range(3))
+
+    stdout, stderr = capfd.readouterr()
+
+    exception_msg = 'Exception while loading results'
+    assert exception_msg not in stdout
+    assert exception_msg not in stderr
+
+
+@with_multiprocessing
+def test_cached_function_race_condition_when_persisting_output_2(tmpdir,
+                                                                 capfd):
+    # Test race condition in first attempt at solving
+    # https://github.com/joblib/joblib/issues/490. The race condition
+    # was due to the delay between seeing the cache directory created
+    # (interpreted as the result being cached) and the output.pkl being
+    # pickled.
+    memory = Memory(cachedir=tmpdir.strpath)
+    func_cached = memory.cache(fast_func_with_conditional_complex_output)
+
+    Parallel(n_jobs=2)(delayed(func_cached)(True if i % 2 == 0 else False)
+                       for i in range(3))
```

Comment:
n_jobs=2 was triggering the race condition every single time in my experience, the reason being that the job 0 is slow to pickle its output, the job 1 is super fast, and the job 2 is seing the cache folder created by the job 0 (interpreted as the result being cached) but not the output.pkl so you have the message with "Exception while trying to load".